python/cudf_polars/cudf_polars/experimental/statistics.py

```diff
@@ -45,13 +46,122 @@ def collect_base_stats(root: IR, config_options: ConfigOptions) -> StatsCollecto
     Returns
     -------
     A new StatsCollector object with populated datasource statistics.
+
+    Notes
+    -----
+    This function initializes the ``StatsCollector`` object
+    with the base datasource statistics. The goal is to build an
+    outline of the statistics that will be collected before any
+    real data is sampled.
     """
     stats: StatsCollector = StatsCollector()
     for node in post_traversal([root]):
+        # Initialize column statistics from datasource information
         stats.column_stats[node] = initialize_column_stats(node, stats, config_options)
+        # Initialize Join-key information
+        initialize_join_key_info(node, stats, config_options)
     return stats
 
 
+def initialize_join_key_info(
+    node: IR, stats: StatsCollector, config_options: ConfigOptions
+) -> None:
+    """
+    Initialize join-key information for the given node.
+
+    Parameters
+    ----------
+    node
+        IR node to initialize join-key information for.
+    stats
+        StatsCollector object to update.
+    config_options
+        GPUEngine configuration options.
+
+    Notes
+    -----
+    This function updates ``stats.joins`` and ``stats.join_keys``.
+    """
+    if isinstance(node, Join):
```

Comment:
My slight style preference would be to push the `isinstance` into `collect_base_stats` and only allow / pass `Join` nodes here to `initialize_join_key_info`. That would make it a bit clearer that `initialize_join_key_info` doesn't do anything sensible for a non-join IR node.